crates/turborepo-lib/src/engine/mod.rs

```diff
@@ -0,0 +1,113 @@
+mod builder;
+
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+};
+
+pub use builder::EngineBuilder;
+use petgraph::Graph;
+
+use crate::{run::task_id::TaskId, task_graph::TaskDefinition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum TaskNode {
+    Root,
+    Task(TaskId<'static>),
+}
+
+impl From<TaskId<'static>> for TaskNode {
+    fn from(value: TaskId<'static>) -> Self {
+        Self::Task(value)
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Building;
+#[derive(Debug, Default)]
+pub struct Built;
+
+#[derive(Debug)]
+pub struct Engine<S = Built> {
+    marker: std::marker::PhantomData<S>,
+    task_graph: Graph<TaskNode, ()>,
+    root_index: petgraph::graph::NodeIndex,
+    task_lookup: HashMap<TaskId<'static>, petgraph::graph::NodeIndex>,
+    task_definitions: HashMap<TaskId<'static>, Rc<TaskDefinition>>,
```

Comment:
We could wrap up these `TaskId` in an `Rc` to prevent multiple allocations, but this introduces the question of where do we dedupe these to avoid multiple `Rc` of the same `TaskId`. Leaving this optimization for future perf work.